server/src/views/profile.js

```diff
@@ -0,0 +1,249 @@
+/* jshint browser:true */
+
+const React = require("react");
+const errors = require("../../shared/errors");
+
+function delay(duration) {
+  return new Promise(resolve => setTimeout(resolve, duration));
+}
+
+exports.ProfileButton = class ProfileButton extends React.Component {
+  constructor(props) {
+    super(props);
+    this.onDocumentClick = this.dismiss.bind(this);
+    this.onRefreshProfileBound = this.onRefreshProfile.bind(this);
+    this.state = {
+      avatarurl: null,
+      nickname: null,
+      email: null,
+      isExpanded: props.initialExpanded,
+      isFetching: false
+    };
+  }
+
+  fetchProfile() {
+    this.setState({
+      isFetching: true,
+      lastError: null
+    }, function () {
+      Promise.race([
+        this.props.getProfile(),
+        delay(5000).then(() => Promise.reject(errors.extTimeout('fetchProfile')))
+      ]).then(
+        profile => {
+          this.setState({
+            isFetching: false,
+            avatarurl: profile && profile.avatarurl,
+            nickname: profile && profile.nickname,
+            email: profile && profile.email
+          });
+        },
+        err => {
+          this.setState({
+            isFetching: false,
+            lastError: err
+          });
+        }
+      ).catch(err => {
+        console.error("Error refreshing profile", err);
+      });
+    });
+  }
+
+  signIn() {
+    Promise.race([
+      this.props.signIn(),
+      delay(2000).then(() => Promise.reject(errors.extTimeout('signIn')))
+    ]).then(() => {
+      this.setState({ isExpanded: false });
+    }, err => {
+      if (err.errno == 1101) {
+        // TODO: Handle extension timeouts.
+        console.error("Timed out fetching profile", err);
+        return;
+      }
+      if (err.errno == 1102) {
+        // User already signed in. Fetch their profile and refresh the view.
+        this.fetchProfile();
+        return;
+      }
+      this.setState({ lastError: err });
+    });
+  }
+
+  signUp() {
+    Promise.race([
+      this.props.signUp(),
+      delay(2000).then(() => Promise.reject(errors.extTimeout('signUp')))
+    ]).then(() => {
+      this.setState({ isExpanded: false });
+    }, err => {
+      if (err.errno == 1101) {
+        // TODO: Timeouts.
+        return;
+      }
+      if (err.errno == 1102) {
+        this.fetchProfile();
+        return;
+      }
+      this.setState({ lastError: err });
+    });
+  }
+
+  updateProfile(profile) {
+    this.props.updateProfile(profile).then(() => {
+      this.fetchProfile();
+    }).catch(err => {
+      console.error("Error updating profile", err);
+    });
+  }
+
+  componentDidMount() {
+    window.addEventListener("click", this.onDocumentClick);
+    document.addEventListener("profile-refresh", this.onRefreshProfileBound);
```

Comment:
If that data is props instead of state, I expect the handler will be global and will trigger a refresh